### DISCUSSION TITLE
Expected Performance For ARM build on MacOS M1

### DISCUSSION THREAD

**Discussion by JeremyLoy** [multiarch/qemu-user-static#195] [Q&A]
hey 👋 

I didn't want to open an issue without first trying out a discussion because I'm not sure if I'm having a problem, or if my expectations just aren't correct.

My goal is to build a docker image containing nodejs on my M1 macbook and deploy it to a Rasberry Pi Zero. So amd64 -> armv6, however it is excruciatingly slow, to the point that I'm not sure if I have something fundamentally broken on my machine.

```shell
docker run --rm -it --platform=linux/arm/v6 alpine /bin/sh
$ uname -m 
> armv7l
$ apk add nodejs npm
$ node --version
> v20.11.1
$ node -e 'console.log("hello world")'
# this never completes. I've tried a few times and its run for nearly a half hour
```

Installing NodeJS through APK was slow, but acceptable. I understand that emulation will be slower, but it was still just a few seconds, maybe a minute to install.

I attached an additional shell session to the running container to measure usage using `top` and was surprised to see such little utilization, however when looking through `Activity Monitor` I was fully maxing out 1 core.

<img width="1005" alt="image" src="https://github.com/multiarch/qemu-user-static/assets/20881794/0f8b0fcd-7b21-4988-b1d9-39a8995cd3eb">

<img width="326" alt="image" src="https://github.com/multiarch/qemu-user-static/assets/20881794/2f28978d-7b92-443f-8fe0-55adb251a0b9">

As for my setup, I'm using `colima` with pretty standard settings. 6 cpu cores allocated. 

I followed the README in this repo to set everything up, so the following command:
```
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

---

So do you all think this is a Node problem? Or do you think there is something wrong with my setup? Do you have anything else I can try to benchmark my system?

As it stands, I might have to just start SCPing my project onto my Rasberry Pi!

Thanks for your time

**Comment by martin-g** [comment]
I'd suggest to `strace` it and see where it "hangs".
```bash
$ strace node -e 'console.log("hello world")'
```

**Comment by JeremyLoy** [comment]
I figured it out. turns out that colima defaults to x86 as the architecture, without rosetta enabled, even if the machine itself is ARM (like the M1 is).

So I was running two layers of poorly optimized emulation!

rebuilding my colima VM with aarch64 as the architecture solved the perf issues, and in general was a speedup across all of my tasks
